tsl/src/partialize_finalize.c

```diff
@@ -260,25 +265,55 @@ inner_agg_deserialize(FACombineFnMeta *combine_meta, bytea *volatile serialized_
 		deser_fcinfo->isnull = false;
 
 		/*
-		 * When an exception is thrown and longjmp() is called, CurrentMemoryContext is potentially
-		 * different than what it was inside the PG_TRY() block below.
+		 * When an exception is thrown and longjmp() is called,
+		 * CurrentMemoryContext is potentially different than what it was
+		 * inside the PG_TRY() block below.
+		 *
+		 * Restore it to the old value so that the code in the subsequent
+		 * PG_CATCH() block does not corrupt the memory.
 		 *
-		 * Restore it to the old value so that the code in the subsequent PG_CATCH() block does not
-		 * corrupt the memory.
+		 * No need for volatile variables since we don't modify any of this
+		 * function's stack frame inside the PG_TRY() block.
 		 *
-		 * No need for volatile variables since we don't modify any of this function's stack frame
-		 * inside the PG_TRY() block.
+		 * We create a subtransaction to run the functions below, in case they
+		 * abort.
 		 */
 		MemoryContext oldcontext = CurrentMemoryContext;
+		ResourceOwner oldowner = CurrentResourceOwner;
+		BeginInternalSubTransaction(NULL);
 		PG_TRY();
 		{
 			deserialized = FunctionCallInvoke(deser_fcinfo);
+			ReleaseCurrentSubTransaction();
+			MemoryContextSwitchTo(oldcontext);
+			CurrentResourceOwner = oldowner;
 		}
 		PG_CATCH();
 		{
-			CurrentMemoryContext = oldcontext;
+			const int sqlerrcode = geterrcode();
+
+			/*
+			 * If we get a data exception category error or a protocol
+			 * violation or an internal error (which is generated by default
+			 * when using elog()), the format is in the old numeric and we can
+			 * attempt to repair it.
+			 *
+			 * If the error is an insufficient resources category we should
+			 * *not* continue executing.
+			 *
+			 * Errors in other categories are not expected, but we cannot do
+			 * anything with them anyway, so just re-throw them too.
+			 */
+			if (ERRCODE_TO_CATEGORY(sqlerrcode) != ERRCODE_DATA_EXCEPTION &&
+				sqlerrcode != ERRCODE_PROTOCOL_VIOLATION && sqlerrcode != ERRCODE_INTERNAL_ERROR)
+			{
+				PG_RE_THROW();
+			}
 			FlushErrorState();
-			/* attempt to repair the serialized partial */
+			RollbackAndReleaseCurrentSubTransaction();
+			MemoryContextSwitchTo(oldcontext);
+			CurrentResourceOwner = oldowner;
+			TS_DEBUG_LOG("attempting repair of serialized partial");
```

Comment:
Suggestion: print memory context name and resource owner here to ensure they are sane after rolling back the error.